### PR TITLE
Api proposition 

### PR DESCRIPTION
**Pull request by mpiniarski**
Proposition of integrating API building functionality to kuick.
`kuick-api` together with `kuick-api-*` modules would provide funtionalities to quickly build flexible APIs in following models: rest, rpc and graphql, providing just minimal required configuration.
Proposed implementations are extensions for ktor framework.

### CODE REVIEW

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        val api: Any,
+        val injector: Injector
+) {
+
+    fun <T> registerRoute(route: RestRoute<T>): Route =
+            parent.route(resourceName, method = route.httpMethod) {
+                println("REST: ${route.httpMethod.value} /$resourceName -> ${route.handler}") // logging
+
+                handle {
+                    val jsonParser = JsonParser()
```
Comment (soywiz) [inline kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: Minor: could this be outside the `handle` block?

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                            route.handler,
+                            jsonParser.parse(call.receiveText()),
+                            call.attributes.allKeys.map { it.name to call.attributes[it as AttributeKey<Any>] }.toMap()
+                    )
+
+                    val result = invokeHandler(api, route.handler, args)
+
+                    val jsonResult = gson.toJsonTree(result)
+
+                    //TODO is it a right place to handle these parameters?
```
Comment (soywiz) [inline kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: As far as I understand you are creating an untyped version of the data (JsonTree) and then you are filtering here before the actual serialization, right? Without heavily modifying toJsonTree to do the filtering, I guess this is the right place. Unless for a critical system, I guess this is a good tradeoff

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+                    if (route.withFieldsParameter && queryParameters.contains("\$include")) {
+                        handleIncludeParam(jsonResult, queryParameters["\$include"]!!, route.includeParameterConfiguration)
+                    }
+
+                    call.respondText(jsonResult.toString(), ContentType.Application.Json) // serialization
+                }
+            }
+
+    private fun handleFieldsParam(jsonResult: JsonElement, fieldsParam: String) {
+        val fieldsParam = (JsonParser().parse(fieldsParam) as JsonArray).map { it.asString }.toHashSet()
```
Comment (soywiz) [inline kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: I guess `toSet` should be enough if you do not plan to mutate it later

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+        fun handleFields(jsonObject: JsonObject) {
+            jsonObject.entrySet().toList().forEach {
```
Comment (soywiz) [inline kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: I bet there is a mutable iterator (with removal support) to do this without cloning collections, if there isn't already a `removeIf`, maybe you create an extension method for it

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+                    }
+                    if (route.withFieldsParameter && queryParameters.contains("\$include")) {
+                        handleIncludeParam(jsonResult, queryParameters["\$include"]!!, route.includeParameterConfiguration)
+                    }
+
+                    call.respondText(jsonResult.toString(), ContentType.Application.Json) // serialization
+                }
+            }
+
+    private fun handleFieldsParam(jsonResult: JsonElement, fieldsParam: String) {
```
Comment (soywiz) [inline kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: Since this acts as a blacklist, what about giving a more specific naming instead of just `handle`?

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        val includeParam = (JsonParser().parse(includeParam) as JsonArray).map { it.asString }.toHashSet()
+
+        fun handleInclude(jsonObject: JsonObject) {
+            jsonObject.entrySet().toList().forEach {
+                if (includeParam.contains(it.key)) {
+                    val method = configuration.getValue(it.key)
+                    jsonObject.remove(it.key)
+                    jsonObject.add(
+                            it.key,
+                            gson.toJsonTree(
```
Comment (soywiz) [inline kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: As far as I understand (correct me if im wrong), this acts as mapper or transformer, right? You can configure a method that transforms a key in the map

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+            }
+        }
+    }
+
+    private fun handleIncludeParam(jsonResult: JsonElement, includeParam: String, configuration: Map<String, (id: String) -> Any>) {
+        val includeParam = (JsonParser().parse(includeParam) as JsonArray).map { it.asString }.toHashSet()
+
+        fun handleInclude(jsonObject: JsonObject) {
+            jsonObject.entrySet().toList().forEach {
+                if (includeParam.contains(it.key)) {
```
Comment (soywiz) [inline kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: Idiomatically this could be `it.key in includeParams`

Comment (soywiz) [reply kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: This also apply to the other method: what if you are serializing a complex object that contains the same field name in other subobjects?

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+                            it.key,
+                            gson.toJsonTree(
+                                    method(it.value.asString)
+                            )
+                    )
+                }
+            }
+        }
+
+        when {
```
Comment (soywiz) [inline kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: This function looks pretty similar to the other one. Could be this structure transformer in a higher order function acting as a json tree visitor?

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+                    handleInclude(obj.asJsonObject)
+                }
+            }
+            else -> {
+            }
+        }
+    }
+}
+
+class RestRoute<T>(
```
Comment (soywiz) [inline kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: Instead of mutating this class an option could be to use a data class, with the fields as parameters with default values, the includeParameters part bool+map could be a nullable map, the with methods could be just calling the copy method of data classes

File: kuick-api-rpc/src/jvmMain/kotlin/kuick/api/rpc/rpc.kt [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+    private fun Any.visitRPC(opAction: (String, KFunction<*>) -> Unit) {
+        //TODO previous version of visitRPC did iterate over interfaces, i'm not sure if it's necessary
+        val srvName = javaClass.simpleName
+        javaClass.kotlin.memberFunctions.forEach { function ->
+            try {
+                opAction(srvName, function)
+            } catch (ieme: Throwable) {
+                println("WARN: invalid public method in controller: $function")
+                ieme.printStackTrace()
```
Comment (soywiz) [inline kuick-api-rpc/src/jvmMain/kotlin/kuick/api/rpc/rpc.kt]: Whats the meaning of ieme?

File: kuick-api/src/jvmMain/kotlin/kuick/api/common.kt [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+import kotlin.coroutines.suspendCoroutine
+import kotlin.reflect.KFunction
+import kotlin.reflect.jvm.javaMethod
+
+
+class IllegalRequestBody : Exception()
+
+suspend fun invokeHandler(api: Any, handler: KFunction<*>, args: Collection<Any?>): Any? =
+        if (handler.isSuspend) {
+            suspendCoroutine { c ->
```
Comment (soywiz) [inline kuick-api/src/jvmMain/kotlin/kuick/api/common.kt]: Why is this needed? You do not seem to be deferring anything

Comment (soywiz) [reply kuick-api/src/jvmMain/kotlin/kuick/api/common.kt]: I guess there is functionality in the kotlin runtime to call functions/methods either suspending or not suspending already without doing it.
Also this method looks pretty generic, so maybe the `api` parameter could be just `instance`, this function called just `invokeMethod` and be placed in another package. Also this might be an extension with the logical receiver KFunction? Isnt there a callSuspend extension or similar?

File: kuick-api/src/jvmMain/kotlin/kuick/api/common.kt [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        else -> throw IllegalRequestBody()
+    }
+    return buildArgs(handler, requestParameters, extraArgs)
+}
+
+
+private fun <T> buildArgs(handler: KFunction<T>,
+                          requestParameters: JsonElement,
+                          extraArgs: Map<String, Any>): Collection<Any?> {
+    val parameters = handler.parameters
```
Comment (soywiz) [inline kuick-api/src/jvmMain/kotlin/kuick/api/common.kt]: I guess this is `.drop(1)`

File: kuick-core/src/jvmMain/kotlin/kuick/reflection/reflection.kt [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+import java.lang.reflect.Modifier
+
+class ReflectionInvocationException(val method: Method, val context: Collection<Any?>, parent: Throwable)
+    : RuntimeException("Error calling method \n  - Method :  [$method\n  - Received context: [${context.joinToString("\n      + ")}]", parent)
+
+
+@Throws(ReflectionInvocationException::class)
+fun Any.invokeWithParams(handlerMethod: Method, params: Collection<Any?>): Any? {
+
+    try {
```
Comment (soywiz) [inline kuick-core/src/jvmMain/kotlin/kuick/reflection/reflection.kt]: Maybe `runCatching`? The catch blocks look pretty similar, maybe `e.targetException ?: e` could unify them

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt (hunk 1 above)
Comment (mpiniarski) [reply kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: Sure. I'm not sure how to handle this JsonParser, cause it's also instantiated again in `handleFieldsParam` and `handleIncludeParam`.
What do you think about adding it as parameter of `Json` object in `kuick-core/src/jvmMain/kotlin/kuick/json/json.kt` next to `gson`?

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt (hunk 2 above)
Comment (mpiniarski) [reply kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: Yes, exactly.

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt (hunk 3 above)
Comment (mpiniarski) [reply kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: Ok

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt (hunk 4 above)
Comment (mpiniarski) [reply kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: There is, thanks ;-)

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt (hunk 6 above)
Comment (mpiniarski) [reply kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: I'm not sure what you mean.
You mean to add a method `JsonObject.transform(key, value)` that will do this replacement?

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt (hunk 7 above)
Comment (mpiniarski) [reply kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: My idea is that these names apply only to first-level object and to handle subobjects dot-separated names have to be provided, e.g.
`include=author,author.avatar` when calling `GET /books` (or `GET /book/{id}`)

This is not handled yet.

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt (hunk 8 above)
Comment (mpiniarski) [reply kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: Done

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt (hunk 5 above)
Comment (mpiniarski) [reply kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: Like `filterFieldsParam`?

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt (hunk 9 above)
Comment (mpiniarski) [reply kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: I've created `RestRoute.Configuration` data class. the thing is that this solution makes dsl code a little more complicated.

I have 2 questions:
1) Why do you thing it's better to use immutable object in this context?
2) Minor: why it's better to use a nullable map instead of bool+map?

File: kuick-api-rpc/src/jvmMain/kotlin/kuick/api/rpc/rpc.kt (hunk 10 above)
Comment (mpiniarski) [reply kuick-api-rpc/src/jvmMain/kotlin/kuick/api/rpc/rpc.kt]: I've copied this part for gokoan actually.
So the truth it that i have do idea ;/ Not sure what to do with this.

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt (hunk 9 above)
Comment (mpiniarski) [reply kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: In the end i wasn't able to implement it with immutable data class. The thing is that in dsl in current implementation we provide a parameter:
`configuration: RestRoute<T>.() -> Unit = {}`
and it makes it very convenient to use `withIncludeParameter` and `withFieldParameter`. If we change it to immutable class, than this parameter would have to be smth like:
`configuration: RestRoute<T>.() -> RestRoute<T> = {RestRoute<T>()}`
and `with*Parameter` methods would have to return RestRoute<T> making them not that convenient to use in DSL, because we would be froced to use them like this:
`withFieldsParameter().withIncludeParameter()`. It looks more like fluent interface, but I don't think it's a good solution for DSL.

File: kuick-api/src/jvmMain/kotlin/kuick/api/common.kt (hunk 11 above)
Comment (mpiniarski) [reply kuick-api/src/jvmMain/kotlin/kuick/api/common.kt]: You're right. I've deleted this code and used `.callSuspend()` instead

File: kuick-core/src/jvmMain/kotlin/kuick/reflection/reflection.kt (hunk 13 above)
Comment (mpiniarski) [reply kuick-core/src/jvmMain/kotlin/kuick/reflection/reflection.kt]: I've deleted this file, I think it's not needed.

File: kuick-api-rest/src/jvmTest/kotlin/kuick/api/rest/Test.kt [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+        intercept(ApplicationCallPipeline.Call) {
+            if (false) {// e.g. if (!ADMIN_USERS.contains(session.userId)) {
+                call.respond(HttpStatusCode.Forbidden)
+                finish()
+            }
+        }
+        build()
+    }
+
+    //TODO Providing additional parameters (later passed to handler method) discuss
```
Comment (mpiniarski) [inline kuick-api-rest/src/jvmTest/kotlin/kuick/api/rest/Test.kt]: We cannot pass additional arguments directy (the ones that are not passed by client directly as parameter but we want to use them when calling handler method e.g. session, userId). Example of passing them directly would be:
```
get(Api:getAll, config = {...}) { handle ->
   withValidSession{ session ->
      handle(extraArgs = listOf(session))
  }
}
```
which on the other hand would force us to write this "withValidSession {...}" for each endpoint separately.

My solution to this problem is to add all those paramters to attributes provided by ktor and than use them when calling handler (https://github.com/NoSpoonLab/kuick/pull/34/files#diff-534e3f0e9a0b54e41a8a610826e9cc46R41).
What do you think about it?

File: kuick-api-rpc/src/jvmMain/kotlin/kuick/api/rpc/rpc.kt [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+import kuick.api.buildArgsFromArray
+import kuick.json.Json.gson
+import kuick.ktor.KuickRouting
+import kotlin.reflect.KFunction
+import kotlin.reflect.full.callSuspend
+import kotlin.reflect.full.memberFunctions
+
+
+// TODO in gokoan rpc handles also downloading and uploading files (see: gokoan/backend/server/src/jvmMain/kotlin/koan/controllers/rpc-controller.kt)
+// TODO discuss: should provide similar functionality here or make solution flexible enough so that user can provide it
```
Comment (mpiniarski) [inline kuick-api-rpc/src/jvmMain/kotlin/kuick/api/rpc/rpc.kt]: The thing is that current solution both for rpc and rest assumes that arguments are passed in path params / query params (not implemented yet) or in request body (either as json array or as json object) and the response is some object serializable to json.
The question is: what to do when parameters are have to be provided in different way e.g. as multipart (which is the case for file upload) or when we want to respond in another way (e.g. download a file).

I'm not sure what to do about it. Methods that handle files should probably not even occurr in REST api. For RPC they might, but I could just assume that it's not a part of this framework responsibility and ignore those cases in infrastructure code. Than user could just define these endpoints on his own.
What do you think?

File: kuick-api-rest/src/jvmTest/kotlin/kuick/api/rest/Test.kt [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+                        field1 = "test1",
+                        field2 = 10
+                )
+        )
+
+        fun getOne(id: String): OtherResource = map[id] ?: throw RuntimeException("404")
+    }
+
+
+    //TODO Pipeline : discuss
```
Comment (mpiniarski) [inline kuick-api-rest/src/jvmTest/kotlin/kuick/api/rest/Test.kt]: I've seen that in `gokoan` in order to this kind of pipeline e.g. for filtering the extension for `PipelineContext<*, ApplicationCall>` are implemented e.g.
```
suspend fun PipelineContext<*, ApplicationCall>.filterHacking(action: suspend () -> Unit) {
    if (HackingMonitor.isHackAttempt(clientIP(), uri())) {
        call.respond(HttpStatusCode.NotFound)
    } else {
        action()
    }
```
The thing is that in this solution we can only apply this filter for one enpoint at a time (it's not a big issue in `gokoan` since all enpoints are generated by infrastrucutre there.
I think that a proper way to do this is by `intercept` mechanism. This way also I don't need to provide any additional mechanism for pipelinening in my solution. In case of using `PipelineContext<*, ApplicationCall>`  extensions i would have to add smth like:
`action: `PipelineContext<*, ApplicationCall>.() -> Unit` to methods in dsl (e.g. `route`, `get`, `put` etc. in rest dsl).

For now there is no way to change the way request in handled in case of rest or rpc APIs other than to use intercept.
I think it's a good solution, but there are 2 things to consider:
1) https://github.com/NoSpoonLab/kuick/pull/34/files#r278100111
2) https://github.com/NoSpoonLab/kuick/pull/34/files#r278106914

File: samples/todo-app2/src/jvmMain/kotlin/kuick/samples/todo2/server.kt [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    installContextPerRequest(injector, DbClientPool { JdbcDriver.connectMemoryH2() }) {
+        injector.getInstance(TodoRepository::class.java).init()
+        injector.getInstance(UserRepository::class.java).init()
+    }
+    installHttpExceptionsSupport()
+
+    kuickRouting {
+        rpcRouting<TodoApi>(injector)
```
Comment (soywiz) [inline samples/todo-app2/src/jvmMain/kotlin/kuick/samples/todo2/server.kt]: Travis seems to be broken at this point:
https://travis-ci.org/NoSpoonLab/kuick/builds/523943661#L868

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt (hunk 5 above)
Comment (soywiz) [reply kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: Yeah, `filter`, `drop` or `remove` would do, I guess. Also the method name should inform about the mutating nature of the function to make it as clear as possible. Verbs usually inform about it, while verbs converted into adjectives are usually better for immutable and pure variants. `withFieldDropped`, `fieldRemoved` vs `dropField`, `removeField`.

I guess that `Param` or `Result` themselves are not relevant for this functionality and can cause confusion.

What about:

`fun JsonElement.removeField`
?

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt (hunk 6 above)
Comment (soywiz) [reply kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: I just wanted to understand what does this do. This whole method applies a function to a field of the JsonElement.

Also ditto about the other method: `jsonResult` -> `element` or `json`. At this point of abstraction `includeParam` would be better understood as `includeField`

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt (hunk 9 above)
Comment (soywiz) [reply kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: Not need for Configuration. Just an initial instance with default values, and then creating copies until it reach the desired state.

I didn't have the whole overview of the implementation so maybe in this case wasn't the best option, not sure.

Regarding to nullability I suggested it for a matter of DRY. Having a boolean and then a potential map with data, that is only used when that boolean has a specific value, is usually represented as a nullable type. In Kotlin nullability is ok, since it is not usually a source of problems.

Rationale `boolean+map -> map?`
Does the map have a non-null value? Let's use it

File: kuick-api-rpc/src/jvmMain/kotlin/kuick/api/rpc/rpc.kt (hunk 15 above)
Comment (soywiz) [reply kuick-api-rpc/src/jvmMain/kotlin/kuick/api/rpc/rpc.kt]: I would ignore this for now. As an initial approach it is good enough.
We probably don't have enough information to do it right at this point.
We would have enough information once it is used and the upload part is required.

File: kuick-api-rest/src/jvmTest/kotlin/kuick/api/rest/Test.kt (hunk 14 above)
Comment (soywiz) [reply kuick-api-rest/src/jvmTest/kotlin/kuick/api/rest/Test.kt]: Yeah, AFAIK Ktor sessions work like that. Even if there was no call access, as long as the method is suspend, we have the coroutine context to place stuff in there like a small per-request dependency container.

File: kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt (hunk 5 above)
Comment (mpiniarski) [reply kuick-api-rest/src/jvmMain/kotlin/kuick/api/rest/rest.kt]: Sounds reasonable.
It would be ```fun JsonElement.preserveFields``` casue this funcionality is whitelist, not blacklist.
What do you think?
